src/clic.adoc

```diff
@@ -2053,6 +2053,199 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+If the Smip extension is implemented, the first and last instructions of a trap handler of an interrupt must be an ipush and the last an ipopxret instruction, respectively.
+If an ipush or ipopxret instruction is used outside of these two scenarios, an illegal-instruction exception shall be raised.
+
+NOTE: In contrast to the Zcmp push/pop intructions, Smip instructions do not encode an incremental stack pointer adjustment in order to ease late preemption and tail chaining.
+
+#TODO: Discuss if we want to have the illegal instruction exceptions as defined here
+
+With these requirements, implementations are allowed to speculatively start executing an ipush instruction in case of a taken interrupt.
+
+#TODO: discuss if we want to enforce this for all interrupts or only HW-vectored interrupts (i.e. coming from xtvt). My assumption is we do not want to do this exceptions
+
+=== Late Preemption of the interrupt handler
+
+While executing the ipush instruction, a hart may optionally check for higher priority/level pending and enabled interrupts at the current privilege mode, and switch to fetching this interrupt handler instead.
+Once the instruction retires, or the interrupts are disabled, no late preemption can occur.
+
+=== Tail chaining of interrupts handlers
+
+While executing the ipopxret instruction, a hart may optionally check for any pending and enabled interrupts at the current privilege mode.
+If such an interrupt exists, the hart may abort executing the ipopxret instruction and trigger the following actions:
+  - Update the xcause.exccode to the corresponding interrupt id.
+  - Perform the trap handler fetch.
+  - Start excuting the trap handler, while skipping the portions of ipush that have not been reverted by the aborted ipopxret.
+
+Explicitly, xepc, xpil, xpp, and xpexccode remain unchanged.
+
+NOTE: In the typical case of tail chaining, the check for another pending enabled interrupt is done after disabling interrupts at the start of the ipopxret instruction. In this case, only the conditional interrupt enablement of the ipush instruction would need to executed.
+
+Once the instruction retires, or the interrupts are disabled, no late preemption can occur.
+
+=== Stack layout
+
+To ease debugging, a stack layout is specified for the saved context.
+
+#TODO: find the best way to specify the memory layout in asciidoc
+
+* xtvec.mode != CLIC:
+** XXLEN = 32:
+
+stack pointer adjustment: 32
+
+24: x15
+
+20: x14
+
+16: x13
+
+12: x12
+
+8: x11
+
+4: x10
+
+** XXLEN = 64:
+
+stack pointer adjustment: 48
+
+48: x15
+
+40: x14
+
+32: x13
+
+24: x12
+
+16: x11
+
+8: x10
+
+* xtvec.mode = CLIC:
+
+start of stackframe remains same. In addition:
+
+** XXLEN = 32
+
+stack pointer adjustment: 32
+
+32: reserved (5) | xpp (2) |  xpie (1) |  xpil (8) | reserved (4) | xpexccode (12)
```

Comment:
This looks almost like `xcause` but not entirely (no `interrupt` bit, and some fields are shuffled slightly) - is this the proposed layout of a new `xpcause` CSR or similar? If so, it would be better to refer to it by this name here, and then specify the register layout and the individual bit fields somewhere else in the spec.

Reply:
the layout is just a strawman. I just want something in place and we can decide later on what the best layout would be.